### ISSUE TITLE
Fatal error - but only when optimized

### ISSUE THREAD

**Issue by kshedstrom** [NOAA-GFDL/MOM6#1201]
After a recent update, my seamount cases are failing with:
`FATAL from PE   3: mpp_sync_self: size_recv does not match of data received`
I used "git bisect" to track it to:
```
commit e8b5d96bcb769b79f68a671e35f4f02940d7cab5
Author: Robert Hallberg <Robert.Hallberg@noaa.gov>
Date:  Wed Aug 19 18:47:35 2020 -0400
```
Running in debug mode to see if I can find out more, it doesn't fail. This is with gnu fortran 8.3.0. I have been unable to get everything to run with the intel compiler, so haven't regularly tried it.

**Comment by Hallberg-NOAA** [comment]
I have tried to reproduce these errors with the dev/gfdl MOM6-examples seamount test cases and the gnu 8.3.0 compiler but have failed to do so. It is possible that this is because our compiler flags or our experiments differed. However, given the nature of the error message, I have some suggestions for how to get at what is going on here.

  The error message you quote indicates that this is a problem with a halo update, and there are two different halo updates associated with this PR. We should be able to identify which is at fault by commenting them out successively. Because the code ran before (albeit incorrectly), it should run again when we comment out the problematic halo update.  The line numbers I will refer to below are those from https://github.com/NOAA-GFDL/MOM6/pull/1184/files. and they might differ from the line numbers in the version of the code you are using.

  The call to `pass_var(fluxes%p_surf,...` on line 1256 of MOM.F90 is only used in coupled models, so it seems to be problematic, but because the test for whether to do this is based on whether a pointer is associated, if we had failed to nullify it at start-up, it might lead to problems. So we should try commenting out this pass var call.

  The other new halo update is done via the group pass at lines 1193-1195.  Commenting out these calls and changing the `elseif (CS%diabatic_first) then` to a simple `if (CS%diabatic_first) then` would restore the previous code.  There are two reasons why this might fail in the way that is described above; the first is if the value of halo_sz returned by extract_diabatic_member differs between processors. The second could be if the group_pass_type variable pass_T_S is not being reset between calls, but instead has inherited inconsistent values from a previous call. To test for the first possibility we could add a line hard coding `halo_sz = 1` before line 1192 . To test the second possibility, we could try using a different group_pass_type variable in place of pass_T_S for the calls at lines 1193-1195. We could also try making the various calls using pass_T_S use consistent message sizes by changing  'To_All+Omit_Corners` to just `To_All` in all cases that use pass_T_S, including those at lines 1199-1201 and perhaps also those at lines 1378-1380 of MOM.F90.

**Comment by kshedstrom** [comment]
Our seamount tests do indeed differ. Mine is an idea of Harper's, with (body-force) tides sloshing past a 3-D seamount and open boundaries.

Thanks for the suggestions! Commenting out line 1256 does not fix it, but adding halo_sz = 1 before line 1192 does.

**Comment by Hallberg-NOAA** [comment]
Thanks for these additional clues.

  After further examining the code, I think that I know what the problem is.  I think that the logical variable CS%Vertex_Shear is not being set in MOM_set_diffusivity.F90 unless CS%useKappaShear is true (i.e., unless USE_JACKSON_PARAM=.true. in your MOM_input file).

  If this is the problem, please delete the line you added before line 1192 of MOM.F90, and then try adding the following line
        `CS%Vertex_Shear = .false.`
just before line 2222 of MOM_set_diffusivity.F90, which reads.
        `if (CS%useKappaShear) CS%Vertex_Shear = kappa_shear_at_vertex(param_file)`
This will ensure that CS%Vertex_Shear is properly initialized in all cases, and can therefore be used safely on line 2237 of the same file.  The reason why this worked with debugging must be that this logical variable was consistently initialized across all the processors, even though it was not formally set.  Curiously, this bug has been here for over 2 years, but is just being tickled now.

  If your seamount test is available from gitHub or elsewhere, please let me know the site so I can try it out myself. Otherwise, if this fix works, please feel free to submit this correction as a PR.

**Comment by kshedstrom** [comment]
Yes, the seamount is available here: https://github.com/ESMG/ESMG-configs/tree/dev/esmg/ocean_only/seamount
Any of layer, sigma, z should work, rho has never worked.

Yes, your fix works.

**Comment by Hallberg-NOAA** [comment]
This issue was corrected by PR#1202, and can now be closed.
